src/Servers/Kestrel/Core/src/Internal/Http/HttpProtocol.cs

```diff
@@ -442,9 +445,26 @@ private void CancelRequestAbortedToken()
         {
             try
             {
-                _abortedCts.Cancel();
-                _abortedCts.Dispose();
-                _abortedCts = null;
+                CancellationTokenSource localAbortCts = null;
+                lock (_abortLock)
+                {
+                    if (_abortedCts != null)
+                    {
+                        if (!_preventRequestAbortedCancellation)
+                        {
+                            localAbortCts = _abortedCts;
+                        }
+                        _abortedCts = null;
+                    }
+                }
+
+                // There is technically a race here where _preventRequestAbortCancellation
+                // changes to true right here and then we call cts.Cancel.
+                // However, this race already existed, where the cts was canceled just before
+                // the _preventRequestAbortedCancellation is set to true.
+                // _preventRequestAbortedCancellation is mostly a heuristic rather than a guarantee
+                // we need to follow.
```

Comment:
This comment is fine for providing an explanation for your design in this PR. Generally speaking though, it's pretty obvious that RequestAborted can fire right before/as _preventRequestAbortedCancellation is set to true.

Reply:
I might add a comment saying that we intentionally don't dispose canceled CTSs though.